components/places/src/db/coop_transaction.rs

```diff
@@ -0,0 +1,159 @@
+///
+/// However, in practice we see the writer thread being starved - even though
+/// it's waiting for a lock, the sync thread still manages to re-get the lock.
+/// In other words, the locks used by sqlite aren't "fair".
+///
+/// So we mitigate this with a simple approach that works fine in these
+/// constraints.
+/// * Each database connection shares a mutex.
+/// * Before starting a transaction, each connection locks the mutex.
+/// * It then starts an "exclusive" transaction - because sqlite now holds a
+///   lock on our behalf, we release the lock on the mutex.
+///
+/// The end result here is that if either connection is waiting for the
+/// database lock because the other already holds it, the waiting one is
+/// guaranteed to get the database lock next.
+impl PlacesDb {
+    pub fn time_chunked_transaction(
+        &self,
+        commit_after: Duration,
+    ) -> Result<TimeChunkedTransaction> {
+        Ok(TimeChunkedTransaction::new(
+            self.conn(),
+            TransactionBehavior::Exclusive,
+            commit_after,
+            &self.coop_tx_lock,
+        )?)
+    }
+
+    pub fn unchecked_transaction(&self) -> Result<UncheckedTransaction> {
+        let _lock = self.coop_tx_lock.lock().unwrap();
+        Ok(UncheckedTransaction::new(
+            self.conn(),
+            TransactionBehavior::Exclusive,
+        )?)
+    }
+}
+
+/// This transaction is suitable for when a transaction is used purely for
+/// performance reasons rather than for data-integrity reasons, or when it's
+/// used for integrity but held longer than strictly necessary for performance
+/// reasons (ie, when it could be multiple transactions and still guarantee
+/// integrity.) Examples of this might be for performance when updating a larger
+/// number of rows, but data integrity concerns could be addressed by using
+/// multiple, smaller transactions.
+///
+/// You can specify a duration for the maximum amount of time the transaction
+/// should be held. You should regularly call .maybe_commit() as part of your
+/// processing, and if the current transaction has been open for greater than
+/// the specified duration the transaction will be committed and another one
+/// started. You should always call .commit() at the end. Note that there is
+/// no .rollback() method as it will be very difficult to work out what was
+/// previously commited and therefore what was rolled back - if you need to
+/// explicitly roll-back, this probably isn't what you should be using. Note
+/// that SQLite might rollback for its own reasons though.
+///
+/// Note that this can still be used for transactions which ensure data
+/// integrity. For example, if you are processing a large group of items, and
+/// each individual item requires multiple updates, you will probably want to
+/// ensure you don't call .maybe_commit() after every item rather than after
+/// each individual database update.
+pub struct TimeChunkedTransaction<'conn> {
```

Comment:
Probably rename this to SequencedTransaction or something.

Reply:
I left it as `TimeChunked` as I think those semantics are also important, but I'm open to another name once you see the new patch.